File: BAEKJOON/2243 - 사탕상자/사탕상자.py
class SegmentTree:
    def __init__(self, n):  # 세그먼트 트리 초기화 메서드
        self.n = n  # 트리에서 관리할 맛의 범위 크기 (예: 1부터 n까지)
        self.tree = [0] * (4 * n)  # 트리 배열 초기화 (4*n 크기)

    def update(self, idx, diff, node, node_left, node_right):  # 트리 업데이트 메서드
        if idx < node_left or node_right < idx:  # 업데이트하려는 인덱스가 현재 노드의 범위 밖이면 종료
            return
        self.tree[node] += diff  # 현재 노드에 diff를 더함 (사탕의 수 변화 반영)
        if node_left != node_right:  # 리프 노드가 아닌 경우
            mid = (node_left + node_right) // 2  # 현재 구간을 반으로 나눔
            # 좌측 자식 노드에 업데이트 전파
            self.update(idx, diff, node * 2, node_left, mid)
            # 우측 자식 노드에 업데이트 전파
            self.update(idx, diff, node * 2 + 1, mid + 1, node_right)

    def query(self, k, node, node_left, node_right):  # k번째 사탕 찾기 위한 쿼리 메서드
        if node_left == node_right:  # 리프 노드에 도달한 경우
            return node_left  # 해당 리프 노드의 인덱스를 반환 (사탕의 맛 번호)
        mid = (node_left + node_right) // 2  # 현재 구간을 반으로 나눔
        if k <= self.tree[node * 2]:  # k번째 사탕이 왼쪽 구간에 있는 경우
            return self.query(k, node * 2, node_left, mid)  # 왼쪽 자식 노드로 이동하여 계속 탐색
        else:  # k번째 사탕이 오른쪽 구간에 있는 경우
            # 오른쪽 자식 노드로 이동하며, 왼쪽 노드의 사탕 수를 k에서 뺌
            return self.query(k - self.tree[node * 2], node * 2 + 1, mid + 1, node_right)

    def add(self, idx, diff):  # 사탕의 수를 추가하거나 제거하는 메서드
        self.update(idx, diff, 1, 1, self.n)  # 인덱스 idx에 해당하는 사탕의 변화를 트리에 반영

    def find_kth(self, k):  # k번째 맛의 사탕을 찾는 메서드
        return self.query(k, 1, 1, self.n)  # k번째 사탕의 맛 번호를 반환

import sys
input = sys.stdin.read  # 입력을 빠르게 처리하기 위해 sys.stdin.read 사용
data = input().splitlines()  # 여러 줄 입력을 받아 줄 단위로 나눔

n = int(data[0])  # 첫 번째 줄: 수정이가 사탕상자에 손을 댄 횟수
commands = [list(map(int, line.split())) for line in data[1:]]  # 나머지 줄: 명령어 리스트로 변환

MAX_TASTE = 1000000  # 사탕의 맛 번호의 최대값 (1부터 1000000까지)
seg_tree = SegmentTree(MAX_TASTE)  # 사탕 맛 범위를 기반으로 세그먼트 트리 생성

result = []  # 출력 결과를 저장할 리스트

for command in commands:  # 각 명령어를 순차적으로 처리
    if command[0] == 1:  # A가 1인 경우 (사탕을 꺼내는 경우)
        k = command[1]  # B: 꺼낼 사탕의 순위
        taste = seg_tree.find_kth(k)  # k번째 사탕의 맛 번호 찾기
        result.append(taste)  # 결과 리스트에 추가
        seg_tree.add(taste, -1)  # 해당 맛의 사탕을 1개 제거
    elif command[0] == 2:  # A가 2인 경우 (사탕을 넣거나 제거하는 경우)
        taste = command[1]  # B: 넣을 사탕의 맛 번호
        count = command[2]  # C: 넣을 사탕의 개수 (음수면 제거)
        seg_tree.add(taste, count)  # 해당 맛의 사탕을 count만큼 추가 또는 제거

sys.stdout.write('\n'.join(map(str, result)) + '\n')  # 모든 출력 결과를 한 번에 출력

File: BAEKJOON/2243 - 사탕상자/사탕상자.cpp
#include <iostream>  // 표준 입출력 사용
#include <vector>    // 벡터 자료구조 사용

using namespace std;

class SegmentTree {
private:
    vector<int> tree;  // 세그먼트 트리를 저장할 벡터
    int size;          // 트리의 크기 (사탕 맛 번호 범위)

    // 트리 업데이트 함수
    void update(int idx, int diff, int node, int nodeLeft, int nodeRight) {
        if (idx < nodeLeft || nodeRight < idx) {  // 업데이트할 인덱스가 범위 밖이면 종료
            return;
        }
        tree[node] += diff;  // 현재 노드에 diff를 더함 (사탕 개수 반영)
        if (nodeLeft != nodeRight) {  // 리프 노드가 아니면 자식 노드로 재귀 호출
            int mid = (nodeLeft + nodeRight) / 2;  // 중간 지점 계산
            update(idx, diff, 2 * node, nodeLeft, mid);        // 왼쪽 자식 노드 업데이트
            update(idx, diff, 2 * node + 1, mid + 1, nodeRight); // 오른쪽 자식 노드 업데이트
        }
    }

    // k번째 사탕의 맛 번호를 찾는 함수
    int query(int k, int node, int nodeLeft, int nodeRight) {
        if (nodeLeft == nodeRight) {  // 리프 노드에 도달한 경우
            return nodeLeft;  // 해당 노드의 인덱스 반환 (사탕 맛 번호)
        }
        int mid = (nodeLeft + nodeRight) / 2;  // 중간 지점 계산
        if (k <= tree[2 * node]) {  // k번째 사탕이 왼쪽 자식 노드에 있는 경우
            return query(k, 2 * node, nodeLeft, mid);  // 왼쪽 자식 노드로 이동
        } else {  // k번째 사탕이 오른쪽 자식 노드에 있는 경우
            return query(k - tree[2 * node], 2 * node + 1, mid + 1, nodeRight);  // 오른쪽 자식 노드로 이동
        }
    }

public:
    SegmentTree(int n) : size(n) {  // 세그먼트 트리 생성자
        tree.resize(4 * n, 0);  // 트리 크기 설정 및 초기화 (0으로 초기화)
    }

    // 사탕 개수를 추가하거나 제거하는 함수
    void add(int idx, int diff) {
        update(idx, diff, 1, 1, size);  // 루트 노드부터 시작하여 트리 업데이트
    }

    // k번째 사탕의 맛 번호를 찾는 함수
    int findKth(int k) {
        return query(k, 1, 1, size);  // 루트 노드부터 시작하여 k번째 사탕 찾기
    }
};

int main() {
    ios::sync_with_stdio(false);  // 입출력 속도 향상을 위한 설정
    cin.tie(nullptr);  // 입출력 동기화를 해제하여 속도 향상

    int n;  // 수정이가 사탕상자에 손을 댄 횟수
    cin >> n;

    SegmentTree segTree(1000000);  // 맛 번호의 최대값 1000000을 갖는 세그먼트 트리 생성

    for (int i = 0; i < n; ++i) {  // n개의 명령어 처리
        int a;  // 명령어 타입 (1: 사탕 꺼내기, 2: 사탕 추가/제거)
        cin >> a;
        if (a == 1) {  // 사탕 꺼내기 명령어 처리
            int b;  // 꺼낼 사탕의 순위
            cin >> b;
            int taste = segTree.findKth(b);  // b번째 사탕의 맛 번호 찾기
            cout << taste << "\n";  // 결과 출력
            segTree.add(taste, -1);  // 해당 맛의 사탕 1개 제거
        } else if (a == 2) {  // 사탕 추가/제거 명령어 처리
            int b, c;  // b: 사탕 맛 번호, c: 사탕 개수 (음수일 경우 제거)
            cin >> b >> c;
            segTree.add(b, c);  // 사탕 개수 업데이트
        }
    }

    return 0;  // 프로그램 종료
}
